_posts/2023-12-18-duckdb-extensions-in-wasm.md

```diff
@@ -0,0 +1,193 @@
+
+To make extension loading work, the DuckDB engine component should block, fetch, validate external WebAssembly code, then plug it in, wiring together import and export, and then the system will be connected and set to keep executing as if it were a single codebase that just gained some new functionality.
+
+
+
+
+The central code block for this to be possible are the following:
+
+
+```c++
+EM_ASM(
+    {
+        const xhr = new XMLHttpRequest();
+        xhr.open("GET", UTF8ToString($0), false);
+        xhr.responseType = "arraybuffer";
+        xhr.send(null);
+        var uInt8Array = xhr.response;
+        // Check signatures / version compatibility left as an exercise
+        WebAssembly.validate(uInt8Array);
+        // Here we add the uInt8Array to Emscripten's filesystem,
+        // for it to be found by dlopen
+        FS.writeFile(UTF8ToString($1), new Uint8Array(uInt8Array));
+    },
+    filename.c_str(), basename.c_str()
+);
+
+auto lib_hdl = dlopen(basename.c_str(), RTLD_NOW | RTLD_LOCAL);
+if (!lib_hdl) {
+    throw IOException(
+      "Extension \"%s\" could not be loaded: %s",
+      filename,
+      GetDLError()
+    );
+}
+```
+
+Here we rely on two powerful feature of [Emscripten](https://emscripten.org/), the compiler toolchain we are using to compile DuckDB to WebAssembly.
+
+First, `EM_ASM` allows to inline JavaScript code directly in C++ code. It means that during runtime when we get to that block of code, the WebAssembly component will go back to JavaScript land, perform a blocking `XMLHttpRequest` on a URL such as [https://extensions.duckdb.org/.../tpch.duckdb_extension.wasm](https://extensions.duckdb.org/duckdb-wasm/v0.9.2/wasm_eh/tpch.duckdb_extension.wasm),
+then validate that what has been just fetched is actually a valid WebAssembly module.
+
+Second, we leverage Emscripten’s [`dlopen` implementation](https://emscripten.org/docs/compiling/Dynamic-Linking.html), which enables compatible WebAssembly modules to be linked together and act as a single composable codebase.
+
+
+<img src="/images/wasm-blog-post-extensions.png"
+     alt="Overview of the architecture of DuckDB-Wasm with extensions"
+     width="828"
+     />
+
+## Developer Guide
+
+We see two main groups of developers using extensions with DuckDB-Wasm.
+
+* Developers working with DuckDB-Wasm: If you are building a website or a library that wraps DuckDB-Wasm, the new extension support means that there is now a wider range of functionality that can be exposed to your users.
+* Developers working on DuckDB extensions: If you have written a DuckDB extension already, or are thinking of doing so, consider porting that also to DuckDB-Wasm. It might be the easiest way to get your extension at the hands of users.
```

Comment:
```suggestion
* Developers working on DuckDB extensions: If you have written a DuckDB extension already, or are thinking of doing so, consider porting that also to DuckDB-Wasm. It might be the easiest way to get your extension in the hands of users.
```